Expose API to update/obfuscate user details in the DB

corinnekrych opened the pull request:
* Please describe what your change is about.
This PR provide a WIT API to obfuscate user and identity information.

* What issues does it solve? (Use [autoreferencing for this](https://help.github.com/articles/basic-writing-and-formatting-syntax/#referencing-issues-and-pull-requests))
[OSIO 804](https://openshift.io/openshiftio/Openshift_io/plan/detail/804)
Background: https://gitlab.cee.redhat.com/dtsd/housekeeping/issues/2108

sbose78 commented on design/account.go:
```diff
@@ -177,6 +177,23 @@ var _ = a.Resource("users", func() {
+		a.Routing(
+			a.PUT("/:id"),
+		)
+		a.Description("Soft delete user for the given ID.")
```
Let's call this "Invalidate user details.."

jarifibrahim commented on account/identity.go:
```diff
@@ -109,6 +110,19 @@ func (m *GormIdentityRepository) Load(ctx context.Context, id uuid.UUID) (*Ident
+		return nil, errs.WithStack(errors.NewNotFoundError("identity", userId.UUID.String()))
+	}
+
+	return &native, errs.WithStack(err)
```
```suggestion
       // Something went wrong with the database query
        if err != nil {
            return nil, errs.Wrapf(err, "failed to load user with ID: %+v", userid)
        }
	return &native, nil
```

jarifibrahim commented on account/identity_blackbox_test.go:
```diff
@@ -86,6 +90,28 @@ func (s *IdentityRepositoryTestSuite) TestOKToDelete() {
+
+func (s *IdentityRepositoryTestSuite) TestOKToObfuscate() {
+	// given
+	identity := createAndLoadUserAndIdentities(s)
```
We have text fixtures for this.
Can you please remove `createAndLoadUserAndIdentities` function?
You can use
```go
fxt := tf.NewTestFixture(s.T(), s.DB, tf.Identities(1), tf.Users(1))
```
You can use `fxt.Identities[0]` and `fxt.Users[0]` in your test.

jarifibrahim commented on account/identity_blackbox_test.go:
```diff
@@ -139,3 +166,37 @@ func createAndLoad(s *IdentityRepositoryTestSuite) *account.Identity {
 	return idnt
 }
+
+func createAndLoadUserAndIdentities(s *IdentityRepositoryTestSuite) *account.Identity {
```
We can use test fixtures to accomplish this.

jarifibrahim commented on design/account.go:
```diff
@@ -177,6 +177,23 @@ var _ = a.Resource("users", func() {
+		a.Response(d.NotFound, JSONAPIErrors)
+		a.Response(d.Unauthorized, JSONAPIErrors)
+		a.Response(d.Forbidden, JSONAPIErrors)
+		a.Response(d.Conflict, JSONAPIErrors)
```
I don't think we need this, do we?

jarifibrahim commented on account/identity.go:
```diff
@@ -80,6 +80,7 @@ type IdentityRepository interface {
 	Lookup(ctx context.Context, username, profileURL, providerType string) (*Identity, error)
 	Save(ctx context.Context, identity *Identity) error
 	Delete(ctx context.Context, id uuid.UUID) error
+	Obfuscate(ctx context.Context, id uuid.UUID, randomStr string) error
```
I agree, let's use `save` method instead of adding a new one.

corinnekrych commented:
@sbose78 @jarifibrahim I've moved all logic in controller in 268a4fa3f0b1c93b6d33a76f07b0d3b9933425a6 but tbh I don't like it, I found the controller more readable with less logic and the `obfuscate` method at DAO level makes sense to me (close to soft delete). 

I've also rebased so let see how the CI checks turn out and if we can merge it.

jarifibrahim commented on account/identity_blackbox_test.go:
```diff
@@ -8,15 +8,16 @@ import (
 	gormtestsupport.DBTestSuite
-	repo account.IdentityRepository
+	repo     account.IdentityRepository
+	repoUser account.UserRepository
```
Is this being used somewhere? It looks like we don't need this.

jarifibrahim commented on account/identity_blackbox_test.go:
```diff
@@ -26,6 +27,7 @@ func TestIdentityRepository(t *testing.T) {
 func (s *IdentityRepositoryTestSuite) SetupTest() {
 	s.DBTestSuite.SetupTest()
 	s.repo = account.NewIdentityRepository(s.DB)
+	s.repoUser = account.NewUserRepository(s.DB)
```
This also can be removed :)

jarifibrahim commented on controller/users.go:
```diff
@@ -49,6 +50,102 @@ func NewUsersController(service *goa.Service, db application.DB, config UsersCon
+	return string(b)
+}
+
+// Obfuscate runs the obfuscate action to soft delete sensitive data associated
```
`to soft delete sensitive data associated` => `to invalidate the sensitive data associated`

jarifibrahim commented on controller/users.go:
```diff
@@ -49,6 +50,102 @@ func NewUsersController(service *goa.Service, db application.DB, config UsersCon
+	if !isSvcAccount {
+		log.Error(ctx, map[string]interface{}{
+			"identity_id": ctx.ID,
+		}, "account used to call create api is not a service account")
```
```suggestion
		}, "account used to call obfuscate API is not a service account")
```

jarifibrahim commented on controller/users.go:
```diff
@@ -49,6 +50,102 @@ func NewUsersController(service *goa.Service, db application.DB, config UsersCon
+		log.Error(ctx, map[string]interface{}{
+			"identity_id": ctx.ID,
+		}, "account used to call create api is not a service account")
+		return jsonapi.JSONErrorResponse(ctx, goa.ErrUnauthorized(errs.New("a non-service account tried to create a user.")))
```
```suggestion
		return jsonapi.JSONErrorResponse(ctx, goa.ErrUnauthorized(errs.New("a non-service account tried to obfuscate a user.")))
```

jarifibrahim commented on controller/users.go:
```diff
@@ -49,6 +50,102 @@ func NewUsersController(service *goa.Service, db application.DB, config UsersCon
+		}, "account used to call create api is not a service account")
+		return jsonapi.JSONErrorResponse(ctx, goa.ErrUnauthorized(errs.New("a non-service account tried to create a user.")))
+	}
+	u, err := uuid.FromString(ctx.ID)
```
Can we use a better variable name, please?
```suggestion
	userID, err := uuid.FromString(ctx.ID)
```

tinakurian replied:
+1

jarifibrahim commented on controller/users.go:
```diff
@@ -49,6 +50,102 @@ func NewUsersController(service *goa.Service, db application.DB, config UsersCon
+		obfStr := randString(12)
+
+		// Obfuscate User
+		users, err := appl.Users().Query(account.UserFilterByID(u))
```
@corinnekrych is there a reason why you chose to use 
```
appl.Users().Query(account.UserFilterByID(u))
```
instead of 
```
appl.Users().Load(ctx, u)
```
?

jarifibrahim commented on controller/users.go:
```diff
@@ -49,6 +50,102 @@ func NewUsersController(service *goa.Service, db application.DB, config UsersCon
+
+		// Obfuscate User
+		users, err := appl.Users().Query(account.UserFilterByID(u))
+		if err != nil || len(users) != 1 {
```
I guess we can get rid of `len(users) != 1` check. We're querying by a `uuid` and you can't get two users with same uuid :)

jarifibrahim replied:
oh, the `len(users) != 1` check makes sense when you use the `appl.Users().Query(..)` method but if you use `appl.Users().Load(..)`  you won't need this check.

jarifibrahim commented on controller/users.go:
```diff
@@ -49,6 +50,102 @@ func NewUsersController(service *goa.Service, db application.DB, config UsersCon
+		return jsonapi.JSONErrorResponse(ctx, goa.ErrBadRequest(errs.New("invalid user id")))
+
+	}
+	returnResponse := application.Transactional(c.db, func(appl application.Application) error {
```
@corinnekrych We'll have to change this. The usual way to use `application.Transactional` block is
```go
err := application.Transactional(c.db, func(appl application.Application) error {
	// do some database stuff
	user, err := appl.Users().Load(ctx, UserID)
	if err != nil {
		return errs.Wrap(err, "failed to load user")
	}
})
// database stuff failed
if err != nil {
	// return error response
}
// return success response
```

The problem with your approach is you're returning errors (like https://github.com/fabric8-services/fabric8-wit/pull/2358/files#diff-da062b312519ee2a61dd669f1ec964d4R112) directly to user (the `returnResponse` will be a database error). We should avoid that :)

jarifibrahim commented on controller/users.go:
```diff
@@ -49,6 +50,102 @@ func NewUsersController(service *goa.Service, db application.DB, config UsersCon
+				"user_id": u,
+				"err":     err,
+			}, "unable to obfuscate the user")
+			return errs.WithStack(err)
```
We should always wrap errors before returning it. It helps in debugging when things go south.
```suggestion
			return errs.Wrap(err, "failed to obfuscate the user")
```

jarifibrahim commented on controller/users.go:
```diff
@@ -49,6 +50,102 @@ func NewUsersController(service *goa.Service, db application.DB, config UsersCon
+			}, "Identity obfuscated!")
+
+		}
+		return ctx.OK([]byte{})
```
This doesn't have to be inside an `appl.Transactional()` block :)

jarifibrahim commented on controller/users_blackbox_test.go:
```diff
@@ -50,6 +50,51 @@ func (s *TestUsersSuite) SecuredServiceAccountController(identity account.Identi
 	return svc, NewUsersController(svc, s.GormDB, s.Configuration)
 }
 
+func (s *TestUsersSuite) TestObfuscateUserAsServiceAccountBadRequest() {
```
@corinnekrych can you please use subtest to structure these tests? I know existing tests aren't structured in a sub-tests manner but it would be great if you could do it.
```go
func (s *TestUsersSuite) TestObfuscateUser(){
	s.T().Run("As a service account", func(t *testing.T){
		t.Run("ok", ...)
		t.Run("bad request", ...)
		t.Run("not found", ...)
	})
	s.T().Run("Not a service account", func(t *testing.T){
		t.Run("unauthorized", ...)
		t.Run("forbidden", ...)
	})
}
```

sbose78 replied:
@corinnekrych  @jarifibrahim I think either way this is fine. With or without subtests :)

jarifibrahim replied:
Yeah. Either way is fine :)

kwk replied:
But that's what subtests are for.

jarifibrahim commented on design/account.go:
```diff
@@ -177,6 +177,22 @@ var _ = a.Resource("users", func() {
+		)
+		a.Description("Invalidate user details for the given userID.")
+		a.Params(func() {
+			a.Param("id", d.String, "id")
```
Can you please change the description and type of param?
```suggestion
			a.Param("userid", d.UUID, "ID of the user whose details are to be invalidated")
```

tinakurian replied:
+1

corinnekrych replied:
the GET/PATCH all use `id` so I'd rather stick with it for consistency

sbose78 replied:
I think `id` is fine because it's anyway being used in the context of users. Also, technically it's identityID ;)

jarifibrahim commented on controller/users.go:
```diff
@@ -49,6 +50,102 @@ func NewUsersController(service *goa.Service, db application.DB, config UsersCon
 		config:     config,
 	}
 }
+func randString(n int) string {
```
```suggestion
func randString(len int) string {
```

tinakurian replied:
+1

jarifibrahim commented on controller/users.go:
```diff
@@ -49,6 +50,102 @@ func NewUsersController(service *goa.Service, db application.DB, config UsersCon
 }
+func randString(n int) string {
+	const letterBytes = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ"
+	b := make([]byte, n)
```
```suggestion
	randomStr := make([]byte, n)
```

tinakurian replied:
+1

corinnekrych commented:
[test]

jarifibrahim commented:
[test]

corinnekrych commented:
@jarifibrahim @sbose78 @tinakurian Good to merge?

kwk commented on controller/users.go:
```diff
@@ -49,6 +50,119 @@ func NewUsersController(service *goa.Service, db application.DB, config UsersCon
+				"user_id": u,
+				"err":     err,
+			}, "unable to load the user")
+			return jsonapi.JSONErrorResponse(ctx, goa.ErrNotFound(err))
```
1. Please use `errors.NewNotFoundError` and not `goa.Err...` ever.
2. You must not use `jsonapi.JSONErrorResponse` within a `application.Transactional` because it will return `ctx.NotFound(jsonErr)` which itself returns `nil` if I'm not mistaken. @xcoulon had raised this PR #1955 explicitly to address this.

corinnekrych commented:
[test]

kwk commented on controller/users.go:
```diff
@@ -94,14 +94,14 @@ func (c *UsersController) Obfuscate(ctx *app.ObfuscateUsersContext) error {
 				"err":     err,
 			}, "unable to find the user")
-			return jsonapi.JSONErrorResponse(ctx, goa.ErrNotFound(errs.New("unable to find the user")))
+			return errors.NewNotFoundError("unable to find the user", u.String())
```
This is the proper way because the message is added automatically: `return errors.NewNotFoundError("user", u.String())`

kwk commented on controller/users.go:
```diff
@@ -130,14 +130,14 @@ func (c *UsersController) Obfuscate(ctx *app.ObfuscateUsersContext) error {
 				"err":     err,
 			}, "unable to retrieve the identity associated to this user id")
-			return jsonapi.JSONErrorResponse(ctx, goa.ErrNotFound(err))
+			return err
```
Don't you wanna return a NotFound error here anymore?

corinnekrych replied:
Here I want to load by the foreign key therefore i use the `Query`, but I'll change to return the NotFoundError

corinnekrych replied:
fixed in b5abde4

sbose78 commented on controller/users_blackbox_test.go:
```diff
@@ -50,6 +50,51 @@ func (s *TestUsersSuite) SecuredServiceAccountController(identity account.Identi
+	// when
+	secureService, secureController := s.SecuredServiceAccountController(identity)
+
+	test.ObfuscateUsersOK(s.T(), secureService.Context, secureService, secureController, (user.ID).String())
```
Some validation missing here :)
Please retrieve the user again, and ensure both the `identity` and the `user` entities have different field values than what they were before the obfuscation was done.

corinnekrych replied:
Indeed, added in commit b5abde4

corinnekrych commented:
[test]

corinnekrych commented:
[test]

sbose78 commented on controller/users.go:
```diff
@@ -49,6 +51,110 @@ func NewUsersController(service *goa.Service, db application.DB, config UsersCon
+		obfStr := randString(12)
+
+		// Obfuscate User
+		user, err := appl.Users().Load(ctx.Context, u)
```
Just use `ctx`, it should work.

sbose78 commented on controller/users.go:
```diff
@@ -49,6 +51,110 @@ func NewUsersController(service *goa.Service, db application.DB, config UsersCon
+		user.URL = obfStr
+		user.ContextInformation = nil
+		user.Company = obfStr
+		if err := appl.Users().Save(ctx.Context, user); err != nil {
```
Just use `ctx`, it should work.

sbose78 commented on controller/users.go:
```diff
@@ -49,6 +51,110 @@ func NewUsersController(service *goa.Service, db application.DB, config UsersCon
+			identity.Username = obfStr
+			identity.ProfileURL = &obfStr
+
+			if err := appl.Identities().Save(ctx.Context, &identity); err != nil {
```
Just use `ctx`, it should work.

corinnekrych replied:
@sbose78 done in 4a81ab1
